// modloader64_api/IModLoaderAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VersionCheckEvent = exports.ModLoaderEvents = void 0;
var ModLoaderEvents;
(function (ModLoaderEvents) {
    ModLoaderEvents["ON_ROM_PATH"] = "ON_ROM_PATH";
    ModLoaderEvents["ON_ROM_HEADER_PARSED"] = "ON_ROM_HEADER_PARSED";
    ModLoaderEvents["ON_CRASH"] = "ON_MODLOADER_CRASH";
    ModLoaderEvents["ON_PRE_ROM_LOAD"] = "ON_PRE_ROM_LOAD";
    ModLoaderEvents["ON_SOFT_RESET_PRE"] = "ON_SOFT_RESET_PRE";
    ModLoaderEvents["ON_SOFT_RESET_POST"] = "ON_SOFT_RESET_POST";
    ModLoaderEvents["ON_ROM_PATCHED_PRE"] = "ON_ROM_PATCHED_PRE";
    ModLoaderEvents["ON_ROM_PATCHED"] = "ON_ROM_PATCHED";
    ModLoaderEvents["ON_ROM_PATCHED_POST"] = "ON_ROM_PATCHED_POST";
    ModLoaderEvents["ON_RECEIVED_CRASH_LOG"] = "ON_RECEIVED_CRASH_LOG";
    ModLoaderEvents["ON_VOLUME_CHANGE"] = "ON_VOLUME_CHANGE";
    ModLoaderEvents["OVERRIDE_TEXTURE_PATH"] = "OVERRIDE_TEXTURE_PATH";
    ModLoaderEvents["ON_EXTERNAL_API_REGISTER"] = "ON_EXTERNAL_API_REGISTER";
})(ModLoaderEvents = exports.ModLoaderEvents || (exports.ModLoaderEvents = {}));
class VersionCheckEvent {
    constructor(ModLoader, plugins) {
        this.canceled = false;
        this.ModLoader = ModLoader;
        this.plugins = plugins;
    }
}
exports.VersionCheckEvent = VersionCheckEvent;
